core/shared/src/main/scala/scalaz/zio/IOQueue.scala

```diff
@@ -38,44 +34,73 @@ class IOQueue[A] private (capacity: Int, ref: IORef[State[A]]) {
    * the queue.
    */
   final def offer[E](a: A): IO[E, Unit] =
-    IO.flatten(ref.modifyFold[E, IO[E, Unit]] {
-      case Deficit(takers) =>
-        takers.dequeueOption match {
-          case None                  => (IO.unit[E], Surplus(Queue.empty[A].enqueue(a), Queue.empty))
-          case Some((taker, takers)) => (taker.complete[E](a).toUnit, Deficit(takers))
-        }
-      case Surplus(values, putters) =>
-        if (values.length < capacity && putters.isEmpty) {
-          (IO.unit[E], Surplus(values.enqueue(a), putters))
-        } else {
-          val p = Promise.unsafeMake[E, Unit]
-          (p.get.ensuring(removePutter(p)), Surplus(values, putters.enqueue((a, p))))
-        }
-    })
+    for {
+      pRef <- IORef[E, Option[Promise[E, Unit]]](None)
+      a <- (for {
+            p <- ref
+                  .modifyFold[Void, (Promise[E, Unit], IO[Void, Unit])] {
+                    case Deficit(takers) =>
+                      val p = Promise.unsafeMake[E, Unit]
+
+                      takers.dequeueOption match {
+                        case None                  => ((p, p.complete(()).toUnit), Surplus(Queue.empty[A].enqueue(a), Queue.empty))
```

Comment:
Since you're discarding the result of the promise, I think you can drop all the spurious `.toUnit` and use `IO.now(false)` on line 56.

Reply:
Yes, it's possible to change to return `IO[Void, Boolean]` and just replace `IO.unit` with `IO.now(false)`. I'm not sure if that's better or not because it may imply to someone reading the code the `Boolean` means something (but it does not).

Reply:
Perhaps `IO[Void, _]` is most explicit to future readers.